File: csharp/Stage2/Evaluator.cs
using System;
using System.Collections.Generic;

namespace MidLang.Stage2
{
    /// <summary>
    /// Evaluator (Interpreter) - Stage 2
    /// 
    /// Purpose: Executes the AST to produce program output.
    /// 
    /// Stage 2 Extensions:
    /// - Symbol table now stores object values (int or string)
    /// - String concatenation with + operator
    /// - Type conversion: int + string converts int to string
    /// - Character literals are treated as single-character strings
    /// </summary>
    public class Evaluator
    {
        /// <summary>
        /// Symbol table: stores variable names and their values.
        /// Values can be int or string (object type).
        /// </summary>
        private readonly Dictionary<string, object> _symbolTable = new Dictionary<string, object>();

        /// <summary>
        /// Evaluates a program by executing all its statements.
        /// </summary>
        public void Evaluate(ProgramNode program)
        {
            foreach (Statement statement in program.Statements)
            {
                EvaluateStatement(statement);
            }
        }

        /// <summary>
        /// Evaluates a statement (var declaration, assignment, or print).
        /// </summary>
        private void EvaluateStatement(Statement statement)
        {
            switch (statement)
            {
                case VarDeclarationStatement varDecl:
                    EvaluateVarDeclaration(varDecl);
                    break;

                case AssignmentStatement assign:
                    EvaluateAssignment(assign);
                    break;

                case PrintStatement print:
                    EvaluatePrint(print);
                    break;

                default:
                    throw new Exception($"Unknown statement type: {statement.GetType()}");
            }
        }

        /// <summary>
        /// Executes a variable declaration: stores the expression's value in the variable.
        /// </summary>
        private void EvaluateVarDeclaration(VarDeclarationStatement varDecl)
        {
            object value = EvaluateExpression(varDecl.Expression);
            _symbolTable[varDecl.VariableName] = value;
        }

        /// <summary>
        /// Executes an assignment statement: stores the expression's value in the variable.
        /// </summary>
        private void EvaluateAssignment(AssignmentStatement assign)
        {
            object value = EvaluateExpression(assign.Expression);
            _symbolTable[assign.VariableName] = value;
        }

        /// <summary>
        /// Executes a print statement: evaluates the expression and outputs the result.
        /// </summary>
        private void EvaluatePrint(PrintStatement print)
        {
            object value = EvaluateExpression(print.Expression);
            Console.WriteLine(value);
        }

        /// <summary>
        /// Evaluates an expression and returns its value (int or string).
        /// </summary>
        private object EvaluateExpression(Expression expression)
        {
            return expression switch
            {
                IntegerLiteral lit => lit.Value,
                StringLiteral lit => lit.Value,
                CharLiteral lit => lit.Value.ToString(), // Convert char to string
                InputIntExpression input => EvaluateInputInt(),
                InputStringExpression input => EvaluateInputString(),
                VariableReference varRef => EvaluateVariable(varRef),
                BinaryExpression binExpr => EvaluateBinaryExpression(binExpr),
                _ => throw new Exception($"Unknown expression type: {expression.GetType()}")
            };
        }

        /// <summary>
        /// Evaluates an inputInt() expression: reads an integer from the console.
        /// </summary>
        private int EvaluateInputInt()
        {
            string input = Console.ReadLine() ?? "";
            if (int.TryParse(input, out int value))
            {
                return value;
            }
            throw new Exception($"Invalid integer input: '{input}'");
        }

        /// <summary>
        /// Evaluates an inputString() expression: reads a string from the console.
        /// </summary>
        private string EvaluateInputString()
        {
            return Console.ReadLine() ?? "";
        }

        /// <summary>
        /// Evaluates a variable reference: looks up the variable's value in the symbol table.
        /// </summary>
        private object EvaluateVariable(VariableReference varRef)
        {
            if (!_symbolTable.ContainsKey(varRef.Name))
            {
                throw new Exception($"Undefined variable: {varRef.Name}");
            }
            return _symbolTable[varRef.Name];
        }

        /// <summary>
        /// Evaluates a binary expression: evaluates left and right, then applies the operator.
        /// Handles:
        /// - Integer arithmetic: +, -, *, /
        /// - String concatenation: + (with automatic int-to-string conversion)
        /// </summary>
        private object EvaluateBinaryExpression(BinaryExpression binExpr)
        {
            object leftObj = EvaluateExpression(binExpr.Left);
            object rightObj = EvaluateExpression(binExpr.Right);

            // String concatenation
            if (binExpr.Operator == "+")
            {
                string leftStr = ConvertToString(leftObj);
                string rightStr = ConvertToString(rightObj);
                return leftStr + rightStr;
            }

            // For other operators, both operands must be integers
            if (!(leftObj is int left) || !(rightObj is int right))
            {
                throw new Exception($"Operator '{binExpr.Operator}' requires integer operands");
            }

            return binExpr.Operator switch
            {
                "-" => left - right,
                "*" => left * right,
                "/" => right == 0 ? throw new Exception("Division by zero") : left / right,
                _ => throw new Exception($"Unknown operator: {binExpr.Operator}")
            };
        }

        /// <summary>
        /// Converts a value to a string representation.
        /// Used for string concatenation and type conversion.
        /// </summary>
        private string ConvertToString(object value)
        {
            return value switch
            {
                int i => i.ToString(),
                string s => s,
                _ => value?.ToString() ?? ""
            };
        }
    }
}

File: cpp/Stage2/Evaluator.cpp
#include "Evaluator.h"
#include <iostream>
#include <stdexcept>
#include <sstream>

void Evaluator::evaluate(ProgramNode* program) {
    for (auto& statement : program->statements) {
        evaluateStatement(statement.get());
    }
}

void Evaluator::evaluateStatement(Statement* statement) {
    if (auto* assign = dynamic_cast<AssignmentStatement*>(statement)) {
        evaluateAssignment(assign);
    } else if (auto* print = dynamic_cast<PrintStatement*>(statement)) {
        evaluatePrint(print);
    } else {
        throw std::runtime_error("Unknown statement type");
    }
}

void Evaluator::evaluateAssignment(AssignmentStatement* assign) {
    auto value = evaluateExpression(assign->expression.get());
    symbolTable[assign->variableName] = value;
}

void Evaluator::evaluatePrint(PrintStatement* print) {
    auto value = evaluateExpression(print->expression.get());
    std::cout << convertToString(value) << std::endl;
}

std::variant<int, std::string> Evaluator::evaluateExpression(Expression* expression) {
    if (auto* lit = dynamic_cast<IntegerLiteral*>(expression)) {
        return lit->value;
    } else if (auto* strLit = dynamic_cast<StringLiteral*>(expression)) {
        return strLit->value;
    } else if (auto* charLit = dynamic_cast<CharLiteral*>(expression)) {
        return std::string(1, charLit->value); // Convert char to string
    } else if (auto* varRef = dynamic_cast<VariableReference*>(expression)) {
        return evaluateVariable(varRef);
    } else if (auto* binExpr = dynamic_cast<BinaryExpression*>(expression)) {
        return evaluateBinaryExpression(binExpr);
    } else {
        throw std::runtime_error("Unknown expression type");
    }
}

std::variant<int, std::string> Evaluator::evaluateVariable(VariableReference* varRef) {
    auto it = symbolTable.find(varRef->name);
    if (it == symbolTable.end()) {
        std::stringstream ss;
        ss << "Undefined variable: " << varRef->name;
        throw std::runtime_error(ss.str());
    }
    return it->second;
}

std::variant<int, std::string> Evaluator::evaluateBinaryExpression(BinaryExpression* binExpr) {
    auto leftObj = evaluateExpression(binExpr->left.get());
    auto rightObj = evaluateExpression(binExpr->right.get());

    // String concatenation
    if (binExpr->op == "+") {
        std::string leftStr = convertToString(leftObj);
        std::string rightStr = convertToString(rightObj);
        return leftStr + rightStr;
    }

    // For other operators, both operands must be integers
    int left, right;
    try {
        left = std::get<int>(leftObj);
        right = std::get<int>(rightObj);
    } catch (const std::bad_variant_access&) {
        std::stringstream ss;
        ss << "Operator '" << binExpr->op << "' requires integer operands";
        throw std::runtime_error(ss.str());
    }

    if (binExpr->op == "-") {
        return left - right;
    } else if (binExpr->op == "*") {
        return left * right;
    } else if (binExpr->op == "/") {
        if (right == 0) {
            throw std::runtime_error("Division by zero");
        }
        return left / right;
    } else {
        std::stringstream ss;
        ss << "Unknown operator: " << binExpr->op;
        throw std::runtime_error(ss.str());
    }
}

std::string Evaluator::convertToString(const std::variant<int, std::string>& value) {
    if (std::holds_alternative<int>(value)) {
        return std::to_string(std::get<int>(value));
    } else {
        return std::get<std::string>(value);
    }
}

